### PR TITLE
Slice api tests

### PR DESCRIPTION
**Pull request by mariuszkubis**

### CODE REVIEW

File: slice-test/slice-core-test/src/test/groovy/com/cognifide/slice/api/qualifier/CurrentResourcePathTest.groovy [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+ * @author Mariusz Kubiś
+ * Date: 10.04.15
+ */
+class CurrentResourcePathTest extends BaseSetup {
+
+    private String path = "/content/testPath/jcr:content"
+
+    def "Current resource path test"() {
+        expect:
+        CurrentResourcePathModel currentResourcePathModel = modelProvider.get(CurrentResourcePathModel.class, path)
```
Comment (mmajchrzak) [inline slice-test/slice-core-test/src/test/groovy/com/cognifide/slice/api/qualifier/CurrentResourcePathTest.groovy]: We should also test it with modelProvider(class, resource)

File: slice-test/slice-core-test/src/test/groovy/com/cognifide/slice/api/tag/SliceTagUtilsTest.groovy [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        bindConfigMethod.invoke(injectorHierarchy, config3)
+
+        //ContextFactory factory = injector2.getInstance(ContextFactory.class)
+        //Context context = factory.getResourceResolverContext(resourceResolver)
+        //contextScope.setContextProvider(new ConstantContextProvider(context))
+
+        //modelProvider = injector.getInstance(ModelProvider.class)
+    }
+
+    def "Has injector"() {
```
Comment (mmajchrzak) [inline slice-test/slice-core-test/src/test/groovy/com/cognifide/slice/api/tag/SliceTagUtilsTest.groovy]: Refactor this. It's nor readable and also seems testing too much as for one test

File: slice-test/slice-mapper-test/src/test/groovy/com/cognifide/slice/mapper/annotation/ChildrenTest.groovy [hunk 3]
```diff
@@ -0,0 +1,73 @@
+package com.cognifide.slice.mapper.annotation
+
+import com.cognifide.slice.test.setup.BaseSetup
+import org.junit.Assert
+
+/**
+ * @author Mariusz Kubiś Date: 14.04.15
+ */
+class ChildrenTest extends BaseSetup {
```
Comment (mmajchrzak) [inline slice-test/slice-mapper-test/src/test/groovy/com/cognifide/slice/mapper/annotation/ChildrenTest.groovy]: This test is only testing happy scenarios. Please add some edge cases, especially if resource is not found or points to not existing resource.

File: slice-test/slice-mapper-test/src/test/groovy/com/cognifide/slice/mapper/annotation/FollowTest.groovy [hunk 4]
```diff
@@ -0,0 +1,42 @@
+package com.cognifide.slice.mapper.annotation
+
+import com.cognifide.slice.test.setup.BaseSetup
+import org.junit.Assert
+
+/**
+ * @author Mariusz Kubiś
+ * Date: 10.04.15
+ */
+class FollowTest extends BaseSetup {
```
Comment (mmajchrzak) [inline slice-test/slice-mapper-test/src/test/groovy/com/cognifide/slice/mapper/annotation/FollowTest.groovy]: Only happy path scenarios covered. Please add edge cases.

File: slice-test/slice-mapper-test/src/test/groovy/com/cognifide/slice/mapper/annotation/IgnorePropertyTest.groovy [hunk 5]
```diff
@@ -0,0 +1,28 @@
+package com.cognifide.slice.mapper.annotation
+
+import com.cognifide.slice.test.setup.BaseSetup
+import org.junit.Assert
+
+/**
+ * @author Mariusz Kubiś Date: 14.04.15
+ */
+class IgnorePropertyTest extends BaseSetup {
```
Comment (mmajchrzak) [inline slice-test/slice-mapper-test/src/test/groovy/com/cognifide/slice/mapper/annotation/IgnorePropertyTest.groovy]: It would be nice to see test for Ignoring properties when:
* various mapping strategies are used
* inheritance is used
* field value is set in code.

File: slice-test/slice-mapper-test/src/test/groovy/com/cognifide/slice/mapper/annotation/InheritanceTest.groovy [hunk 6]
```diff
@@ -0,0 +1,26 @@
+package com.cognifide.slice.mapper.annotation
+
+import com.cognifide.slice.test.setup.BaseSetup
+import org.junit.Assert
+
+/**
+ * @author Mariusz Kubiś Date: 15.04.15
+ */
+class InheritanceTest extends BaseSetup {
```
Comment (mmajchrzak) [inline slice-test/slice-mapper-test/src/test/groovy/com/cognifide/slice/mapper/annotation/InheritanceTest.groovy]: Please add following test cases:
* various mapping strategies are used (both on parent and child classes, all combinations should be covered)
* @SliceResource applied on parent and (not) on child class

File: slice-test/slice-mapper-test/src/test/groovy/com/cognifide/slice/mapper/annotation/MappingStrategyTest.groovy [hunk 7]
```diff
@@ -0,0 +1,37 @@
+package com.cognifide.slice.mapper.annotation
+
+import com.cognifide.slice.test.setup.BaseSetup
+import org.junit.Assert
+
+/**
+ * @author Mariusz Kubiś Date: 15.04.15
+ */
+class MappingStrategyTest extends BaseSetup {
```
Comment (mmajchrzak) [inline slice-test/slice-mapper-test/src/test/groovy/com/cognifide/slice/mapper/annotation/MappingStrategyTest.groovy]: Please add fields with set values and check how they are overridden by Slice depending on strategy

File: slice-test/slice-core-test/src/test/groovy/com/cognifide/slice/api/qualifier/CurrentResourcePathTest.groovy (hunk 1 above)
Comment (mariuszkubis) [inline slice-test/slice-core-test/src/test/groovy/com/cognifide/slice/api/qualifier/CurrentResourcePathTest.groovy]: done

File: slice-test/slice-mapper-test/src/test/groovy/com/cognifide/slice/mapper/annotation/ChildrenTest.groovy (hunk 3 above)
Comment (mariuszkubis) [inline slice-test/slice-mapper-test/src/test/groovy/com/cognifide/slice/mapper/annotation/ChildrenTest.groovy]: done

File: slice-core/pom.xml [hunk 8]
```diff
@@ -143,6 +143,12 @@
 			<groupId>org.apache.sling</groupId>
 			<artifactId>org.apache.sling.scripting.api</artifactId>
 		</dependency>
+
+		<dependency>
+			<groupId>com.google.guava</groupId>
+			<artifactId>guava</artifactId>
+			<version>18.0</version>
```
Comment (mmajchrzak) [inline slice-core/pom.xml]: It's used for tests only so add test scope

File: slice-core/src/main/java/com/cognifide/slice/core/internal/module/AnnotationReader.java [hunk 9]
```diff
@@ -58,22 +58,19 @@ public AnnotationVisitor visitAnnotation(String paramString, boolean paramBoolea
...
-		return result;
+		String name = annotation.getName();
+		return annotations.contains(name);
 	}
 
 	private String getAnnotationClassName(String rawName) {
 		return rawName.substring(1, rawName.length() - 1).replace('/', '.');
 	}
 
+	public List<String> getAnnotations() {
```
Comment (mmajchrzak) [inline slice-core/src/main/java/com/cognifide/slice/core/internal/module/AnnotationReader.java]: package visibility would be more appropriate. But to be honest I'd just test this AnnotationReader as a black box using its isAnnotationPresent method only (for verification).

File: slice-test/slice-mapper-test/src/test/groovy/com/cognifide/slice/mapper/annotation/FollowTest.groovy [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+	def "Follow test with null value"() {
+		setup: "Create a test content, under path '/content/bar', with property 'jcrPropertyModel' set to null"
+		pageBuilder.content {
+			bar("cq:PageContent") { "jcr:content"("jcrPropertyModel": null) }
+		}
+		when: "Getting Follow model, based on null path"
+		modelProvider.get(FollowModel.class, "/content/bar/jcr:content")
+
+		then: "ProvisionException should be thrown - resource can not be null"
+		thrown(ProvisionException)
```
Comment (mmajchrzak) [inline slice-test/slice-mapper-test/src/test/groovy/com/cognifide/slice/mapper/annotation/FollowTest.groovy]: Boom! You don't expect Slice to throw an exception here. Because it does, then we've found a bug ;)

File: slice-test/slice-mapper-test/src/test/groovy/com/cognifide/slice/mapper/annotation/FollowTest.groovy [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+			bar("cq:PageContent") { "jcr:content"("jcrPropertyModel": true) }
+		}
+		when: "Getting Follow model, based on non-string property"
+		modelProvider.get(FollowModel.class, "/content/bar/jcr:content")
+
+		then: "ProvisionException should be thrown - resource can not be null"
+		thrown(ProvisionException)
+	}
+
+	@FailsWith(ProvisionException)
```
Comment (mmajchrzak) [inline slice-test/slice-mapper-test/src/test/groovy/com/cognifide/slice/mapper/annotation/FollowTest.groovy]: Same here, and in the following test methods

File: slice-test/slice-mapper-test/src/test/groovy/com/cognifide/slice/mapper/annotation/InheritanceTest.groovy [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+
+	def "Inheritance test: Setup test content"() {
```
Comment (mmajchrzak) [inline slice-test/slice-mapper-test/src/test/groovy/com/cognifide/slice/mapper/annotation/InheritanceTest.groovy]: Hmm - this test is testing prosper itself. We don't need it - please remove
